### PR TITLE
Update grpc to 1.29.1

### PR DESCRIPTION
**Pull request by hevrard**
Needed to fix gcc-9.2.1-31 compilation issue:

external/com_github_grpc_grpc/src/core/lib/iomgr/ev_epollex_linux.cc:1105:13: error: ambiguating new declaration of 'long int gettid()'

The fix on grpc side is: https://github.com/grpc/grpc/commit/de6255941a5e1c2fb2d50e57f84e38c09f45023d
Update to the latest grpc to get that fix in.

### CODE REVIEW

**Comment by hevrard** [comment]
Fix #60

File: gapir/cc/grpc_replay_service.h [hunk 1]
```diff
@@ -39,8 +37,8 @@ class ReplayResponse;
 namespace gapir {
 
 using ReplayGrpcStream =
-    grpc::ServerReaderWriter<replay_service::ReplayResponse,
-                             replay_service::ReplayRequest>;
+    grpc_impl::ServerReaderWriter<replay_service::ReplayResponse,
```
Comment (pau-baiget) [inline gapir/cc/grpc_replay_service.h:40]: Why is this needed? In the fix description you talk about upgrading the gRPC version but i don't see why you need to call the _impl directly here.

Comment (purvisa-at-google-com) [reply gapir/cc/grpc_replay_service.h:40]: Guess this goes for the changed include/declaration above too.

Comment (hevrard) [reply gapir/cc/grpc_replay_service.h:40]: This change is needed to be able to compile, the ServerReaderWriter template has been renamed in the new version.

Basically all the changes of this PR are there to be able to pass build with the new grpc version.

**Comment by hevrard** [comment]
I'm not sure why the build fails on mac/win.

@pau-baiget could you try to build this PR on mac & windows?

File: WORKSPACE [hunk 2]
```diff
@@ -33,5 +33,11 @@ load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
 go_rules_dependencies()
 go_register_toolchains()
 
+load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps", "grpc_test_only_deps")
```
Comment (pau-baiget) [inline WORKSPACE:36]: And why is this needed if it wasn't needed before? Just curiosity.

Comment (pau-baiget) [reply WORKSPACE:36]: Actually something more than curiosity: the build failures in Win/Mac are found in a dependency of grpc, probably loaded here.

**Comment by tobine** [comment]
I verified that this fixes the build fail I was seeing on Linux

File: WORKSPACE (hunk 2 above)
Comment (hevrard) [reply WORKSPACE:36]: I had to add these to resolve grpc dependency issues.

See: https://github.com/grpc/grpc/blob/560577371abff8f504091d906ddaf58126a07762/bazel/grpc_extra_deps.bzl#L10-L27

But I guess we should be able to remove the `grpc_test_only_deps`, can you try this?

Comment (silence-do-good) [reply WORKSPACE:36]: I checked it locally and it works. The build seems to finish successfully without the test_only_deps

Comment (silence-do-good) [reply WORKSPACE:36]: However, I have only a linux machine where I can build. For some reason, bazel doesn't run on my Mac, so I'm not able to verify on any other platform.

Comment (hevrard) [reply WORKSPACE]: @silence-do-good thanks for checking, I removed the `grpc_test_only_deps`

Comment (hevrard) [reply WORKSPACE]: @pau-baiget I also added the link above as a comment.

**Comment by hevrard** [comment]
This is basically blocked on https://github.com/abseil/abseil-cpp/issues/326.

Recap:
- this grpc update is motivated to compile with a new version of gcc9
- the update transitively depends on abseil which has an open issue that prevents to build on macOS: https://github.com/abseil/abseil-cpp/issues/326
- to work around this, other projects patch abseil as part of their build (e.g. tensorflow, see https://github.com/tensorflow/tensorflow/commit/04e169ac74a81b00507f1bb571513a5c2f61547f); we don't want to go this way
- we decide to use gcc8 on all builds for now, see b/157873397

Closing, we may reuse this once GRPC depends on a version of Abseil that has a fix for the macOS builds.

**Comment by pau-baiget** [comment]
Added a couple commits to patch the dependency (abseil-cpp) that's causing the build failure on macOS. The reason we're patching instead of forcing gcc-8 everywhere is that in Linux we can't enforce an older glibc (the problems happen from glibc 2.30). Once the https://github.com/abseil/abseil-cpp/issues/326 is fixed then we can remove the patch.

**Comment by hevrard** [comment]
#319 is our fix for now. This PR here might come handy when we eventually want to upgrade grpc.
